File: core/pipeline.py
import concurrent.futures
from core.financials import fetch_stock_financials, safe_float
from core.scorer import compute_brs

def run_pipeline(universe: list[dict]) -> dict:
    stages = {
        "after_stage1": len(universe),
        "after_stage2": 0,
        "after_stage3": 0,
        "after_stage4": 0,
        "after_stage5": 0,
        "after_stage6": 0,
    }

    symbols = [s["symbol"] for s in universe]

    financials_map = {}
    with concurrent.futures.ThreadPoolExecutor(max_workers=10) as executor:
        futures = {executor.submit(fetch_stock_financials, sym): sym for sym in symbols}
        for future in concurrent.futures.as_completed(futures):
            sym = futures[future]
            try:
                financials_map[sym] = future.result()
            except Exception as e:
                print(f"[pipeline] {sym} fetch failed: {e}")

    candidates = []

    for stock in universe:
        sym = stock["symbol"]
        fin = financials_map.get(sym)
        if not fin:
            continue

        signals = {
            "earnings_inflection": False,
            "opm_expanding": False,
            "debt_declining": False,
            "positive_cash_flow": False,
            "roce_inflecting": False,
            "promoter_stable": False,
            "institutional_entry": False,
            "price_mismatch": False,
        }
        metrics = {}

        # ── Stage 2: Earnings Inflection ──────────────────────────
        pat_raw = list(reversed(fin.get("pat_by_q", [])))
        pat = [v for v in pat_raw if v != 0.0]

        revenue_raw = list(reversed(fin.get("revenue_by_q", [])))
        revenue = [v for v in revenue_raw if v != 0.0]

        opm_raw = list(reversed(fin.get("opm_by_q", [])))
        opm = [v for v in opm_raw if v != 0.0]

        if len(pat) >= 6:
            try:
                pat_growth_latest = ((pat[-1] - pat[-5]) / abs(pat[-5])) * 100 if pat[-5] != 0 else 0
                pat_growth_prev = ((pat[-2] - pat[-6]) / abs(pat[-6])) * 100 if pat[-6] != 0 else 0
                metrics["pat_growth_latest_q"] = safe_float(pat_growth_latest)
                metrics["pat_growth_prev_q"] = safe_float(pat_growth_prev)

                if pat_growth_latest > 20:
                    signals["earnings_inflection"] = True
                    if pat_growth_prev > 10 and pat_growth_latest > pat_growth_prev:
                        metrics["earnings_accelerating"] = True
                    else:
                        metrics["earnings_accelerating"] = False
            except Exception as e:
                print(f"[pipeline] {sym} Stage 2 math error: {e}")

        elif len(pat) >= 5:
            try:
                pat_growth_latest = ((pat[-1] - pat[-5]) / abs(pat[-5])) * 100 if pat[-5] != 0 else 0
                metrics["pat_growth_latest_q"] = safe_float(pat_growth_latest)
                metrics["pat_growth_prev_q"] = 0.0
                if pat_growth_latest > 20:
                    signals["earnings_inflection"] = True
            except Exception as e:
                print(f"[pipeline] {sym} Stage 2 fallback error: {e}")

        if not signals["earnings_inflection"]:
            continue

        stages["after_stage2"] += 1

        # OPM expansion bonus
        if len(opm) >= 2:
            metrics["opm_latest"] = safe_float(opm[-1])
            metrics["opm_prev"] = safe_float(opm[-2])
            if opm[-1] - opm[-2] > 1.5:
                signals["opm_expanding"] = True

        # ── Stage 3: Balance Sheet ─────────────────────────────────
        de_raw = list(reversed(fin.get("debt_to_equity_by_year", [])))
        de = [v for v in de_raw if v != 0.0]

        ocf_raw = list(reversed(fin.get("operating_cf_by_year", [])))
        ocf = [v for v in ocf_raw if v != 0.0]

        if len(de) >= 2:
            metrics["debt_to_equity"] = safe_float(de[-1])
            metrics["debt_to_equity_prev"] = safe_float(de[-2])
            if de[-1] < de[-2]:
                signals["debt_declining"] = True

        if len(ocf) >= 1 and ocf[-1] > 0:
            signals["positive_cash_flow"] = True
            metrics["operating_cf"] = safe_float(ocf[-1])

        if not (signals["debt_declining"] or signals["positive_cash_flow"]):
            continue

        stages["after_stage3"] += 1

        # ── Stage 4: ROCE Inflection ───────────────────────────────
        roce_raw = list(reversed(fin.get("roce_by_year", [])))
        roce = [v for v in roce_raw if v != 0.0]

        if len(roce) >= 2:
            metrics["roce_3yr"] = roce[-min(3, len(roce)):]
            if roce[-1] > roce[-2] and roce[-1] > 8:
                signals["roce_inflecting"] = True
            elif roce[-1] > 20:
                signals["roce_inflecting"] = True

        if not signals["roce_inflecting"]:
            continue

        stages["after_stage4"] += 1

        # ── Stage 5: Institutional / Promoter ─────────────────────
        signals["promoter_stable"] = True
        metrics["promoter_holding_change"] = 0.0
        metrics["fii_change"] = 0.0
        stages["after_stage5"] += 1

        # ── Stage 6: Price-Earnings Mismatch ──────────────────────
        pat_year_raw = list(reversed(fin.get("pat_by_year", [])))
        pat_year = [v for v in pat_year_raw if v != 0.0]

        earnings_growth_3yr = 0.0
        if len(pat_year) >= 4 and pat_year[-4] > 0 and pat_year[-1] > 0:
            try:
                earnings_growth_3yr = ((pat_year[-1] / pat_year[-4]) ** (1/3) - 1) * 100
            except:
                pass
        elif len(pat_year) >= 3 and pat_year[-3] > 0 and pat_year[-1] > 0:
            try:
                earnings_growth_3yr = ((pat_year[-1] / pat_year[-3]) ** (1/2) - 1) * 100
            except:
                pass
        elif len(pat_year) >= 2 and pat_year[-2] > 0 and pat_year[-1] > 0:
            try:
                earnings_growth_3yr = ((pat_year[-1] / pat_year[-2]) - 1) * 100
            except:
                pass

        price_return_1yr = safe_float(fin.get("price_return_1yr", 0))
        mismatch_score = earnings_growth_3yr - price_return_1yr

        metrics["earnings_growth_3yr"] = safe_float(earnings_growth_3yr)
        metrics["price_return_1yr"] = safe_float(price_return_1yr)
        metrics["mismatch_score"] = safe_float(mismatch_score)

        if mismatch_score > 5:
            signals["price_mismatch"] = True

        stages["after_stage6"] += 1

        brs = compute_brs(signals, metrics)

        candidates.append({
            "symbol": sym,
            "name": fin.get("name", sym),
            "current_price": safe_float(fin.get("current_price", 0)),
            "market_cap": safe_float(fin.get("market_cap", 0)),
            "pe": safe_float(fin.get("pe", 0)),
            "roe": safe_float(fin.get("roe", 0)),
            "brs": brs,
            "signals": signals,
            "metrics": metrics,
            "screener_url": f"https://www.screener.in/company/{sym}/",
            "tv_url": f"https://www.tradingview.com/chart/?symbol=NSE:{sym}",
        })

    candidates.sort(key=lambda x: x["brs"], reverse=True)
    return {"candidates": candidates, "stages": stages}

File: lib/pipeline.ts
import type { TVStock } from './tradingview.js';
import type { ParsedFinancials, BreakoutCandidate } from './types.js';
import { parseFinancials } from './screener.js';
import { fetchPriceReturn1yr } from './tradingview.js';
import { computeBRS } from './scorer.js';

function getLast(arr: number[], offset: number = 0): number {
  const idx = arr.length - 1 - offset;
  return idx >= 0 ? arr[idx] : 0;
}

function computeGrowth(current: number, previous: number): number {
  if (previous === 0) return 0;
  return ((current - previous) / Math.abs(previous)) * 100;
}

interface StockData {
  stock: TVStock;
  financials: ParsedFinancials;
}

export async function runPipeline(
  universe: TVStock[],
  options: { includeSME: boolean; onProgress?: (msg: string) => void }
): Promise<{ candidates: BreakoutCandidate[]; stages: Record<string, number> }> {
  const stages: Record<string, number> = {};
  stages.afterStage1 = universe.length;
  options.onProgress?.('Stage 1: Universe fetched with ' + universe.length + ' stocks');

  const batchSize = 10;
  const batchDelayMs = 2000;
  const financialsMap = new Map<string, ParsedFinancials | null>();

  for (let i = 0; i < universe.length; i += batchSize) {
    const batch = universe.slice(i, i + batchSize);
    const results = await Promise.allSettled(
      batch.map(s => parseFinancials(s.nseSymbol))
    );
    results.forEach((r, idx) => {
      const symbol = batch[idx].nseSymbol;
      if (r.status === 'fulfilled') {
        financialsMap.set(symbol, r.value);
      } else {
        console.warn(`Failed to fetch financials for ${symbol}:`, r.reason);
        financialsMap.set(symbol, null);
      }
    });
    if (i + batchSize < universe.length) {
      await new Promise(r => setTimeout(r, batchDelayMs));
    }
  }

  options.onProgress?.('Stage 2: Financials fetched for all stocks');

  const stage2Passed: StockData[] = [];
  for (const stock of universe) {
    const fin = financialsMap.get(stock.nseSymbol);
    if (!fin) continue;

    const patByQ = fin.patByQ;
    if (patByQ.length < 6) {
      console.warn(`Skipping ${stock.nseSymbol} — insufficient quarterly PAT data (${patByQ.length} quarters)`);
      continue;
    }

    const last = patByQ.length - 1;
    const patGrowthLatestQ = computeGrowth(patByQ[last], patByQ[last - 4]);
    const patGrowthPrevQ = computeGrowth(patByQ[last - 1], patByQ[last - 5]);

    if (patGrowthLatestQ > 25 && patGrowthPrevQ > 15 && patGrowthLatestQ > patGrowthPrevQ) {
      stage2Passed.push({ stock, financials: fin });
    }
  }
  stages.afterStage2 = stage2Passed.length;
  options.onProgress?.(`Stage 2: Earnings inflection — ${stage2Passed.length} passed`);

  const stage3Passed: StockData[] = [];
  for (const sd of stage2Passed) {
    const fin = sd.financials;
    const dte = fin.debtToEquityByYear;
    const ocf = fin.operatingCFByYear;

    const lastDte = dte.length - 1;
    const debtDeclining = dte.length >= 3
      && dte[lastDte] < dte[lastDte - 1]
      && dte[lastDte - 1] < dte[lastDte - 2];

    const lastOcf = ocf.length - 1;
    const positiveCashFlow = ocf.length >= 2
      && ocf[lastOcf] > 0
      && ocf[lastOcf - 1] > 0;

    if (debtDeclining || positiveCashFlow) {
      stage3Passed.push(sd);
    }
  }
  stages.afterStage3 = stage3Passed.length;
  options.onProgress?.(`Stage 3: Balance sheet quality — ${stage3Passed.length} passed`);

  const stage4Passed: StockData[] = [];
  for (const sd of stage3Passed) {
    const fin = sd.financials;
    const roce = fin.roceByYear;
    if (roce.length < 3) continue;

    const last = roce.length - 1;
    const roceInflecting = roce[last] > roce[last - 2] && roce[last] > 8;

    if (roceInflecting) {
      stage4Passed.push(sd);
    }
  }
  stages.afterStage4 = stage4Passed.length;
  options.onProgress?.(`Stage 4: ROCE inflection — ${stage4Passed.length} passed`);

  const stage5Passed: StockData[] = [];
  for (const sd of stage4Passed) {
    const fin = sd.financials;
    const promoter = fin.promoterByQ;
    if (promoter.length < 5) {
      stage5Passed.push(sd);
      continue;
    }

    const last = promoter.length - 1;
    const promoterHoldingChange = promoter[last] - promoter[last - 4];
    const promoterStable = promoterHoldingChange > -1.0;

    if (promoterStable) {
      stage5Passed.push(sd);
    }
  }
  stages.afterStage5 = stage5Passed.length;
  options.onProgress?.(`Stage 5: Institutional discovery — ${stage5Passed.length} passed`);

  const tvSymbols = stage5Passed.map(sd => sd.stock.symbol);
  const priceReturnMap = await fetchPriceReturn1yr(tvSymbols);

  const stage6Passed: BreakoutCandidate[] = [];
  for (const sd of stage5Passed) {
    const fin = sd.financials;
    const patByYear = fin.patByYear;
    if (patByYear.length < 4) continue;

    const last = patByYear.length - 1;
    const earningsGrowth3yr = computeGrowth(patByYear[last], patByYear[last - 3]);

    const priceReturn1yr = priceReturnMap.get(sd.stock.nseSymbol) ?? 0;
    const mismatchScore = earningsGrowth3yr - priceReturn1yr;

    if (mismatchScore > 5) {
      const roce = fin.roceByYear;
      const dte = fin.debtToEquityByYear;
      const ocf = fin.operatingCFByYear;

      const patGrowthLatestQ = computeGrowth(
        getLast(fin.patByQ, 0),
        getLast(fin.patByQ, 4)
      );
      const patGrowthPrevQ = computeGrowth(
        getLast(fin.patByQ, 1),
        getLast(fin.patByQ, 5)
      );

      const opmByQ = fin.opmByQ;
      const opmLatest = opmByQ.length > 0 ? getLast(opmByQ, 0) : 0;
      const opmPrev = opmByQ.length > 1 ? getLast(opmByQ, 1) : 0;

      const debtToEquity = dte.length > 0 ? getLast(dte, 0) : 0;
      const debtToEquityPrevYear = dte.length > 1 ? getLast(dte, 1) : 0;

      const operatingCashFlow = ocf.length > 0 ? getLast(ocf, 0) : 0;

      const lastDte = dte.length - 1;
      const debtDeclining = dte.length >= 3
        && dte[lastDte] < dte[lastDte - 1]
        && dte[lastDte - 1] < dte[lastDte - 2];

      const lastOcf = ocf.length - 1;
      const positiveCashFlow = ocf.length >= 2
        && ocf[lastOcf] > 0
        && ocf[lastOcf - 1] > 0;

      const roceLast = roce.length - 1;
      const roceInflecting = roce.length >= 3
        && roce[roceLast] > roce[roceLast - 2]
        && roce[roceLast] > 8;

      const promoter = fin.promoterByQ;
      const promoterHoldingChange = promoter.length >= 5
        ? promoter[promoter.length - 1] - promoter[promoter.length - 5]
        : 0;
      const promoterStable = promoterHoldingChange > -1.0;

      const fii = fin.fiiByQ;
      const fiiChange = fii.length >= 2
        ? fii[fii.length - 1] - fii[fii.length - 2]
        : 0;
      const institutionalEntry = fiiChange > 0.1;

      const dii = fin.diiByQ;
      const mfCountChange = dii.length >= 5
        ? (dii[dii.length - 1] > 0 ? 1 : 0) - (dii[dii.length - 5] > 0 ? 1 : 0)
        : 0;

      const signals = {
        earningsInflection: true,
        opmExpanding: opmByQ.length >= 2 && (opmByQ[opmByQ.length - 1] - opmByQ[opmByQ.length - 2]) > 1.5,
        debtDeclining,
        positiveCashFlow,
        roceInflecting,
        promoterStable,
        institutionalEntry,
        priceMismatch: true,
      };

      const roce3yr = roce.length >= 3
        ? [roce[roce.length - 3], roce[roce.length - 2], roce[roce.length - 1]]
        : [...roce];

      const metrics = {
        patGrowthLatestQ,
        patGrowthPrevQ,
        opmLatest,
        opmPrev,
        debtToEquity,
        debtToEquityPrevYear,
        operatingCashFlow,
        roce3yr,
        promoterHoldingChange,
        mfCountChange,
        fiiChange,
        earningsGrowth3yr,
        priceReturn1yr,
        mismatchScore,
      };

      const brs = computeBRS(signals, metrics);

      const candidate: BreakoutCandidate = {
        symbol: sd.stock.nseSymbol,
        name: sd.stock.name,
        currentPrice: sd.stock.price,
        marketCap: sd.stock.marketCap,
        pe: sd.stock.pe,
        roe: sd.stock.roe,
        roce: roce.length > 0 ? getLast(roce, 0) : 0,
        brs,
        signals,
        metrics,
        screenerUrl: `https://www.screener.in/company/${sd.stock.nseSymbol}/`,
        tvSymbol: sd.stock.symbol,
        fetchedAt: new Date().toISOString(),
      };

      stage6Passed.push(candidate);
    }
  }
  stages.afterStage6 = stage6Passed.length;
  options.onProgress?.(`Stage 6: Price-earnings mismatch — ${stage6Passed.length} passed`);

  stage6Passed.sort((a, b) => b.brs - a.brs);

  return { candidates: stage6Passed, stages };
}
